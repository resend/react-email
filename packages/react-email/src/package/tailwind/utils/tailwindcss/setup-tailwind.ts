import type { Rule } from 'patched-postcss';
import parse from 'patched-postcss/lib/parse';
import collapseAdjacentRules from 'patched-tailwindcss/lib/lib/collapseAdjacentRules';
import collapseDuplicateDeclarations from 'patched-tailwindcss/lib/lib/collapseDuplicateDeclarations';
import evaluateTailwindFunctions from 'patched-tailwindcss/lib/lib/evaluateTailwindFunctions';
import expandApplyAtRules from 'patched-tailwindcss/lib/lib/expandApplyAtRules';
import expandTailwindAtRules from 'patched-tailwindcss/lib/lib/expandTailwindAtRules';
import { generateRules as rawGenerateRules } from 'patched-tailwindcss/lib/lib/generateRules';
import partitionApplyAtRules from 'patched-tailwindcss/lib/lib/partitionApplyAtRules';
import resolveDefaultsAtRules from 'patched-tailwindcss/lib/lib/resolveDefaultsAtRules';
import substituteScreenAtRules from 'patched-tailwindcss/lib/lib/substituteScreenAtRules';
import type { TailwindConfig } from '../../tailwind';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { setupTailwindContext } from './setup-tailwind-context';

const tailwindAtRulesRoot = parse(
  `
  @tailwind base;
  @tailwind components;
`,
).root();

export function setupTailwind(config: TailwindConfig) {
  const tailwindContext = setupTailwindContext(config);
  return {
    generateRootForClasses: (classes: string[]) => {
      const bigIntRuleTuples: [bigint, Rule][] = rawGenerateRules(
        new Set(classes),
        tailwindContext,
      );

      const root = tailwindAtRulesRoot
        .clone()
        .append(...bigIntRuleTuples.map(([, rule]) => rule));
      partitionApplyAtRules()(root);
      // This is fine because the internal await is never actually called out
      // because of there not being any `changedContent` files on the context
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      expandTailwindAtRules(tailwindContext)(root);
      partitionApplyAtRules()(root);
      expandApplyAtRules(tailwindContext)(root);
      evaluateTailwindFunctions(tailwindContext)(root);
      substituteScreenAtRules(tailwindContext)(root);
      resolveDefaultsAtRules(tailwindContext)(root);
      collapseAdjacentRules(tailwindContext)(root);
      collapseDuplicateDeclarations(tailwindContext)(root);

      resolveAllCSSVariables(root);

      return root;
    },
  };
}
