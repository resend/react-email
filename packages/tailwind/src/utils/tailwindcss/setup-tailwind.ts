import { type CssNode, parse } from 'css-tree';
import { type Config, compile } from 'tailwindcss';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { resolveCalcExpressions } from '../css/resolve-calc-expressions';
import { sanitizeDeclarations } from '../css/sanitize-declarations';
import { sanitizeNonInlinableRules } from '../css/sanitize-non-inlinable-rules';
import indexCss from './tailwind-stylesheets/index';
import preflightCss from './tailwind-stylesheets/preflight';
import themeCss from './tailwind-stylesheets/theme';
import utilitiesCss from './tailwind-stylesheets/utilities';

export type TailwindSetup = Awaited<ReturnType<typeof setupTailwind>>;

export async function setupTailwind(config: Config) {
  const baseCss = `
@layer theme, base, components, utilities;
@import "tailwindcss/theme.css" layer(theme);
@import "tailwindcss/utilities.css" layer(utilities);
@config;
`;
  const compiler = await compile(baseCss, {
    async loadModule(id, base, resourceHint) {
      if (resourceHint === 'config') {
        return {
          path: id,
          base: base,
          module: config,
        };
      }

      throw new Error(
        `NO-OP: should we implement support for ${resourceHint}?`,
      );
    },
    polyfills: 3, // All
    async loadStylesheet(id, base) {
      if (id === 'tailwindcss') {
        return {
          base,
          path: 'tailwindcss/index.css',
          content: indexCss,
        };
      }

      if (id === 'tailwindcss/preflight.css') {
        return {
          base,
          path: id,
          content: preflightCss,
        };
      }

      if (id === 'tailwindcss/theme.css') {
        return {
          base,
          path: id,
          content: themeCss,
        };
      }

      if (id === 'tailwindcss/utilities.css') {
        return {
          base,
          path: id,
          content: utilitiesCss,
        };
      }

      throw new Error(
        'stylesheet not supported, you can only import tailwindcss',
      );
    },
  });

  let addedUtilities: string[] = [];

  let root: CssNode | undefined;

  return {
    /**
     * @description Given a list of Tailwind classes, it generates the corresponding CSS. Also resolves simple `calc` functions, and css variables.
     * The returned CSS also includes the CSS generated from all previous calls to this function, this is internal to Tailwind.
     */
    addUtilities: function addUtilities(
      classes: string[],
      { compatibilityFixes = true } = {},
    ) {
      if (!compatibilityFixes) {
        return parse(compiler.build(classes));
      }

      if (classes.some((className) => !addedUtilities.includes(className))) {
        addedUtilities = [...addedUtilities, ...classes];
        root = parse(compiler.build(classes));
        resolveAllCSSVariables(root);
        resolveCalcExpressions(root);
        sanitizeDeclarations(root);
        sanitizeNonInlinableRules(root);
      }
      return root!;
    },
  };
}
