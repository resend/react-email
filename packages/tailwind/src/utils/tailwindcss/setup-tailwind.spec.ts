import { generate } from 'css-tree';
import { setupTailwind } from './setup-tailwind';

test('setupTailwind() and addUtilities()', async () => {
  const { addUtilities } = await setupTailwind({});

  expect(
    generate(addUtilities(['text-red-500', 'sm:bg-blue-300', 'bg-slate-900'])),
  ).toMatchSnapshot();

  expect(generate(addUtilities(['bg-red-100']))).toMatchSnapshot();
});
