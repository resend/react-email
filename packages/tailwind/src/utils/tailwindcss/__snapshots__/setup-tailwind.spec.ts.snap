// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setupTailwind() and addUtilities() 1`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer utilities{.bg-slate-900{background-color:rgb(14.6,23,43)}.text-red-500{color:rgb(250.6,43.8,54.3)}.sm_bg-blue-300{@media (width>=40rem){background-color:rgb(141.9,197.2,255)!important}}}"`;

exports[`setupTailwind() and addUtilities() 2`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer utilities{.bg-red-100{background-color:rgb(254.8,225.8,225.8)}.bg-slate-900{background-color:rgb(14.6,23,43)}.text-red-500{color:rgb(250.6,43.8,54.3)}.sm_bg-blue-300{@media (width>=40rem){background-color:rgb(141.9,197.2,255)!important}}}"`;
